frontend/src/test/java/com/google/tripmeout/frontend/servlet/ServletUtilTest.java

```diff
@@ -0,0 +1,152 @@
+
+    Optional<List> extractedList = ServletUtil.extractFromRequestBody(request, gson, List.class);
+    assertThat(extractedList).isPresent();
+    assertThat(extractedList.get()).containsExactly("hello", "my", "name", "is");
+  }
+
+  @Test
+  public void extractFromRequestBody_supplyTripObject_returnsOptionalTrip() throws IOException {
+    when(request.getReader())
+        .thenReturn(new BufferedReader(new StringReader(
+            "{id: abc123, name: trip1, userId: a, locationLat: 33.2, locationLong: -22.77}")));
+
+    TripModel testTrip = TripModel.builder()
+                             .setId("abc123")
+                             .setName("trip1")
+                             .setUserId("a")
+                             .setLocationLat(33.2)
+                             .setLocationLong(-22.77)
+                             .build();
+
+    assertThat(ServletUtil.extractFromRequestBody(request, gson, TripModel.class))
+        .hasValue(testTrip);
+  }
+
+  @Test
+  public void extractFromRequestBody_supplyTripObjectNotAllFields_returnsOptionalTrip()
+      throws IOException {
+    when(request.getReader())
+        .thenReturn(new BufferedReader(new StringReader("{id: abc123, userId: a}")));
+
+    TripModel testTrip = TripModel.builder().setId("abc123").setUserId("a").build();
+
+    assertThat(ServletUtil.extractFromRequestBody(request, gson, TripModel.class))
+        .hasValue(testTrip);
+  }
+
+  @Test
+  public void extractFromRequestBody_missingId_returnsOptionalTrip() throws IOException {
+    when(request.getReader())
+        .thenReturn(new BufferedReader(
+            new StringReader("{name: trip1, userId: a, locationLat: 33.2, locationLong: -22.77}")));
+    assertThat(ServletUtil.extractFromRequestBody(request, gson, TripModel.class)).isEmpty();
+  }
+
+  @Test
+  public void extractFromRequestBody_supplyPlaceVisitObject_returnsOptionalPlaceVisitModel()
+      throws IOException {
+    when(request.getReader())
+        .thenReturn(new BufferedReader(new StringReader(
+            "{tripId: abc123, placeId: 123, name: trip1, userMark: YES, latitude: 33.2, longitude: -22.77}")));
+
+    PlaceVisitModel testPlace1 = PlaceVisitModel.builder()
+                                     .setTripId("abc123")
+                                     .setPlaceId("123")
+                                     .setName("trip1")
+                                     .setUserMark(PlaceVisitModel.UserMark.YES)
+                                     .setLatitude(33.2)
+                                     .setLongitude(-22.77)
+                                     .build();
+
+    assertThat(ServletUtil.extractFromRequestBody(request, gson, PlaceVisitModel.class))
+        .hasValue(testPlace1);
+  }
+
+  @Test
+  public void parseUri_noMatchesTripNamePattern_throwsError() {
+    when(request.getRequestURI()).thenReturn("http://tripmeout.com/trip/abc123");
```

Comment:
nit: `request::getRequestUri` would never return something like this. It always returns the path specified after the host name and before the query params.

if the http request is `http://tripmeout.com/trip/abc123?q=random` `request::getRequestUri` returns `/trip/abc123`